src/core/code-actions/fix-code-actions-provider.ts

```diff
@@ -0,0 +1,100 @@
+import { CodeAction, CodeActionKind, TextDocument, Range, languages, WorkspaceEdit, commands } from 'vscode';
+import { BaseCodeActionsProvider, CodeActionContextExtended, DiagnosticExtended, ValidationResultExtended } from './base-code-actions-provider';
+import { Fix, Replacement } from 'sarif';
+import { COMMANDS } from '../../constants';
+import globals from '../../utils/globals';
+import { raiseWarning } from '../../utils/errors';
+import { trackEvent } from '../../utils/telemetry';
+
+
+class FixCodeActionsProvider extends BaseCodeActionsProvider<FixCodeAction> {
+  private _warning: Promise<unknown> | null;
+
+  public async provideCodeActions(document: TextDocument, _range: Range, context: CodeActionContextExtended) {
+    return this.getMonokleDiagnostics(context).filter((diagnostic: DiagnosticExtended) => diagnostic.result?.fixes?.length > 0).map((diagnostic: DiagnosticExtended) => {
+      return new FixCodeAction(document, diagnostic);
+    });
+  }
+
+  public async resolveCodeAction(codeAction: FixCodeAction) {
+    const user = await globals.getUser();
+
+    if (!user.isAuthenticated) {
+      if (!this._warning) {
+        this._warning = raiseWarning(`Fix requires authentication.`, [{
+          title: 'Login',
+          callback() {
+            commands.executeCommand(COMMANDS.LOGIN);
+          },
+        }]).then(() => this._warning = null);
+
+        trackEvent('code_action/fix', {
+          status: 'cancelled',
+          error: 'Unauthenticated'
+        });
+      }
```

Comment:
I wonder if this should be here? Because then it will run when you hover over a quick fix, right?

IMHO we should run it when user tries to apply the action. This can be done by adding dedicated command and attaching it to CodeAction only when user is not logged in, or we can add a command which just runs any arbitrary function which is passed to id (via `command.arguments`).  WDYT?

Reply:
ebff3afd946d2342c7008c38d6e20a681f0c5db8